File: plugin-sdk/typescript/src/api/debug.ts
/**
 * Debug API
 *
 * @description Development and debugging utilities for plugin development.
 * Provides tracing, breakpoints, performance monitoring, and logging.
 */

/**
 * Log level for debug messages
 */
export type DebugLogLevel = 'trace' | 'debug' | 'info' | 'warn' | 'error';

/**
 * Debug log entry
 */
export interface DebugLogEntry {
  /** Log level */
  level: DebugLogLevel;
  /** Log message */
  message: string;
  /** Timestamp */
  timestamp: number;
  /** Additional data */
  data?: unknown;
  /** Source location (file:line) */
  source?: string;
  /** Stack trace if available */
  stack?: string;
}

/**
 * Performance trace entry
 */
export interface TraceEntry {
  /** Trace name/label */
  name: string;
  /** Start time (performance.now()) */
  startTime: number;
  /** End time (performance.now()) */
  endTime?: number;
  /** Duration in milliseconds */
  duration?: number;
  /** Nested traces */
  children?: TraceEntry[];
  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Performance metrics
 */
export interface PerformanceMetrics {
  /** Plugin ID */
  pluginId: string;
  /** Total execution time (ms) */
  totalTime: number;
  /** Number of tool calls */
  toolCallCount: number;
  /** Average tool call duration (ms) */
  avgToolCallDuration: number;
  /** Number of hook invocations */
  hookInvocationCount: number;
  /** Average hook duration (ms) */
  avgHookDuration: number;
  /** Memory usage (bytes) */
  memoryUsage?: number;
  /** All trace entries */
  traces: TraceEntry[];
}

/**
 * Breakpoint definition
 */
export interface Breakpoint {
  /** Breakpoint ID */
  id: string;
  /** Type of breakpoint */
  type: 'hook' | 'tool' | 'event' | 'custom';
  /** Target name (hook name, tool name, etc.) */
  target: string;
  /** Condition expression (optional) */
  condition?: string;
  /** Whether breakpoint is enabled */
  enabled: boolean;
  /** Hit count */
  hitCount: number;
}

/**
 * Debug session state
 */
export interface DebugSession {
  /** Session ID */
  id: string;
  /** Plugin ID */
  pluginId: string;
  /** Session start time */
  startedAt: Date;
  /** Whether session is active */
  active: boolean;
  /** Current breakpoints */
  breakpoints: Breakpoint[];
  /** Captured logs */
  logs: DebugLogEntry[];
  /** Performance metrics */
  metrics: PerformanceMetrics;
}

/**
 * Slow operation alert
 */
export interface SlowOperation {
  /** Operation type */
  type: 'tool' | 'hook' | 'ipc' | 'custom';
  /** Operation name */
  name: string;
  /** Duration in milliseconds */
  duration: number;
  /** Threshold that was exceeded */
  threshold: number;
  /** Timestamp */
  timestamp: number;
  /** Additional context */
  context?: Record<string, unknown>;
}

/**
 * Debug API for plugin development
 *
 * @remarks
 * Provides development tools for debugging, performance monitoring,
 * and troubleshooting plugins during development.
 *
 * @example
 * ```typescript
 * // Start a debug session
 * context.debug.startSession();
 *
 * // Add a breakpoint
 * context.debug.setBreakpoint('hook', 'onAgentStep');
 *
 * // Trace performance
 * const endTrace = context.debug.startTrace('myOperation');
 * await doSomething();
 * endTrace();
 *
 * // Measure a function
 * const result = await context.debug.measure('fetchData', async () => {
 *   return await fetchData();
 * });
 *
 * // Get metrics
 * const metrics = context.debug.getMetrics();
 * console.log('Total time:', metrics.totalTime);
 *
 * // Monitor slow operations
 * context.debug.onSlowOperation(1000, (op) => {
 *   console.warn(`Slow operation: ${op.name} took ${op.duration}ms`);
 * });
 * ```
 */
export interface PluginDebugAPI {
  /**
   * Start a debug session
   *
   * @returns Session ID
   */
  startSession(): string;

  /**
   * End the current debug session
   */
  endSession(): void;

  /**
   * Get current debug session
   *
   * @returns Current session or null
   */
  getSession(): DebugSession | null;

  /**
   * Check if debug mode is active
   *
   * @returns Whether debug mode is active
   */
  isActive(): boolean;

  /**
   * Set a breakpoint
   *
   * @param type - Breakpoint type
   * @param target - Target name
   * @param condition - Optional condition expression
   * @returns Breakpoint ID
   */
  setBreakpoint(
    type: Breakpoint['type'],
    target: string,
    condition?: string
  ): string;

  /**
   * Remove a breakpoint
   *
   * @param breakpointId - Breakpoint ID to remove
   */
  removeBreakpoint(breakpointId: string): void;

  /**
   * Enable/disable a breakpoint
   *
   * @param breakpointId - Breakpoint ID
   * @param enabled - Whether to enable or disable
   */
  toggleBreakpoint(breakpointId: string, enabled: boolean): void;

  /**
   * Get all breakpoints
   *
   * @returns Array of breakpoints
   */
  getBreakpoints(): Breakpoint[];

  /**
   * Clear all breakpoints
   */
  clearBreakpoints(): void;

  /**
   * Start a performance trace
   *
   * @param name - Trace name/label
   * @param metadata - Optional metadata
   * @returns Function to end the trace
   */
  startTrace(name: string, metadata?: Record<string, unknown>): () => void;

  /**
   * Measure execution time of a function
   *
   * @param name - Measurement name
   * @param fn - Function to measure
   * @returns Result of the function
   */
  measure<T>(name: string, fn: () => T | Promise<T>): Promise<T>;

  /**
   * Get performance metrics
   *
   * @returns Performance metrics for this plugin
   */
  getMetrics(): PerformanceMetrics;

  /**
   * Reset performance metrics
   */
  resetMetrics(): void;

  /**
   * Log a debug message
   *
   * @param level - Log level
   * @param message - Log message
   * @param data - Optional additional data
   */
  log(level: DebugLogLevel, message: string, data?: unknown): void;

  /**
   * Get captured logs
   *
   * @param level - Optional level filter
   * @param limit - Maximum number of logs to return
   * @returns Array of log entries
   */
  getLogs(level?: DebugLogLevel, limit?: number): DebugLogEntry[];

  /**
   * Clear captured logs
   */
  clearLogs(): void;

  /**
   * Monitor for slow operations
   *
   * @param threshold - Duration threshold in ms
   * @param handler - Handler called when slow operation is detected
   * @returns Unsubscribe function
   */
  onSlowOperation(
    threshold: number,
    handler: (operation: SlowOperation) => void
  ): () => void;

  /**
   * Capture a snapshot of current state
   *
   * @param label - Snapshot label
   * @returns Snapshot data
   */
  captureSnapshot(label: string): {
    label: string;
    timestamp: number;
    metrics: PerformanceMetrics;
    logs: DebugLogEntry[];
  };

  /**
   * Export debug data for analysis
   *
   * @returns Debug data as JSON string
   */
  exportData(): string;

  /**
   * Assert a condition (throws in debug mode)
   *
   * @param condition - Condition to assert
   * @param message - Error message if assertion fails
   */
  assert(condition: boolean, message: string): void;

  /**
   * Time a block of code with console.time-like API
   *
   * @param label - Timer label
   */
  time(label: string): void;

  /**
   * End a timer and log the duration
   *
   * @param label - Timer label
   */
  timeEnd(label: string): void;

  /**
   * Log current memory usage
   */
  logMemory(): void;
}

File: plugin-sdk/python/cognia/debug.py
"""
Debug API for Cognia Plugin SDK

Provides development and debugging utilities for plugin development.
Includes tracing, breakpoints, performance monitoring, and logging.
"""

from abc import ABC, abstractmethod
from typing import Any, Callable, Dict, List, Optional
from dataclasses import dataclass

from .types import (
    DebugLogLevel,
    DebugLogEntry,
    TraceEntry,
    PerformanceMetrics,
    Breakpoint,
    DebugSession,
    SlowOperation,
)


class DebugAPI(ABC):
    """
    Debug API for plugin development.
    
    Provides development tools for debugging, performance monitoring,
    and troubleshooting plugins during development.
    
    Example:
        # Start a debug session
        context.debug.start_session()
        
        # Add a breakpoint
        context.debug.set_breakpoint('hook', 'on_agent_step')
        
        # Trace performance
        end_trace = context.debug.start_trace('my_operation')
        await do_something()
        end_trace()
        
        # Measure a function
        result = await context.debug.measure('fetch_data', fetch_data)
        
        # Get metrics
        metrics = context.debug.get_metrics()
        print('Total time:', metrics.total_time)
    """
    
    @abstractmethod
    def start_session(self) -> str:
        """
        Start a debug session.
        
        Returns:
            Session ID
        """
        pass
    
    @abstractmethod
    def end_session(self) -> None:
        """End the current debug session"""
        pass
    
    @abstractmethod
    def get_session(self) -> Optional[DebugSession]:
        """
        Get current debug session.
        
        Returns:
            Current session or None
        """
        pass
    
    @abstractmethod
    def is_active(self) -> bool:
        """
        Check if debug mode is active.
        
        Returns:
            Whether debug mode is active
        """
        pass
    
    @abstractmethod
    def set_breakpoint(self, bp_type: str, target: str, condition: Optional[str] = None) -> str:
        """
        Set a breakpoint.
        
        Args:
            bp_type: Breakpoint type ('hook', 'tool', 'event', 'custom')
            target: Target name
            condition: Optional condition expression
            
        Returns:
            Breakpoint ID
        """
        pass
    
    @abstractmethod
    def remove_breakpoint(self, breakpoint_id: str) -> None:
        """
        Remove a breakpoint.
        
        Args:
            breakpoint_id: Breakpoint ID to remove
        """
        pass
    
    @abstractmethod
    def toggle_breakpoint(self, breakpoint_id: str, enabled: bool) -> None:
        """
        Enable/disable a breakpoint.
        
        Args:
            breakpoint_id: Breakpoint ID
            enabled: Whether to enable or disable
        """
        pass
    
    @abstractmethod
    def get_breakpoints(self) -> List[Breakpoint]:
        """
        Get all breakpoints.
        
        Returns:
            Array of breakpoints
        """
        pass
    
    @abstractmethod
    def clear_breakpoints(self) -> None:
        """Clear all breakpoints"""
        pass
    
    @abstractmethod
    def start_trace(self, name: str, metadata: Optional[Dict[str, Any]] = None) -> Callable[[], None]:
        """
        Start a performance trace.
        
        Args:
            name: Trace name/label
            metadata: Optional metadata
            
        Returns:
            Function to end the trace
        """
        pass
    
    @abstractmethod
    async def measure(self, name: str, fn: Callable) -> Any:
        """
        Measure execution time of a function.
        
        Args:
            name: Measurement name
            fn: Function to measure
            
        Returns:
            Result of the function
        """
        pass
    
    @abstractmethod
    def get_metrics(self) -> PerformanceMetrics:
        """
        Get performance metrics.
        
        Returns:
            Performance metrics for this plugin
        """
        pass
    
    @abstractmethod
    def reset_metrics(self) -> None:
        """Reset performance metrics"""
        pass
    
    @abstractmethod
    def log(self, level: DebugLogLevel, message: str, data: Optional[Any] = None) -> None:
        """
        Log a debug message.
        
        Args:
            level: Log level
            message: Log message
            data: Optional additional data
        """
        pass
    
    @abstractmethod
    def get_logs(self, level: Optional[DebugLogLevel] = None, limit: Optional[int] = None) -> List[DebugLogEntry]:
        """
        Get captured logs.
        
        Args:
            level: Optional level filter
            limit: Maximum number of logs to return
            
        Returns:
            Array of log entries
        """
        pass
    
    @abstractmethod
    def clear_logs(self) -> None:
        """Clear captured logs"""
        pass
    
    @abstractmethod
    def on_slow_operation(self, threshold: float, handler: Callable[[SlowOperation], None]) -> Callable[[], None]:
        """
        Monitor for slow operations.
        
        Args:
            threshold: Duration threshold in ms
            handler: Handler called when slow operation is detected
            
        Returns:
            Unsubscribe function
        """
        pass
    
    @abstractmethod
    def capture_snapshot(self, label: str) -> Dict[str, Any]:
        """
        Capture a snapshot of current state.
        
        Args:
            label: Snapshot label
            
        Returns:
            Snapshot data with label, timestamp, metrics, and logs
        """
        pass
    
    @abstractmethod
    def export_data(self) -> str:
        """
        Export debug data for analysis.
        
        Returns:
            Debug data as JSON string
        """
        pass
    
    @abstractmethod
    def assert_condition(self, condition: bool, message: str) -> None:
        """
        Assert a condition (throws in debug mode).
        
        Args:
            condition: Condition to assert
            message: Error message if assertion fails
        """
        pass
    
    @abstractmethod
    def time(self, label: str) -> None:
        """
        Time a block of code with console.time-like API.
        
        Args:
            label: Timer label
        """
        pass
    
    @abstractmethod
    def time_end(self, label: str) -> None:
        """
        End a timer and log the duration.
        
        Args:
            label: Timer label
        """
        pass
    
    @abstractmethod
    def log_memory(self) -> None:
        """Log current memory usage"""
        pass


__all__ = [
    "DebugAPI",
]
